src/server/cluster/cluster_config.h

```diff
@@ -3,36 +3,73 @@
 //
 
 #pragma once
-#include <absl/container/flat_hash_set.h>
 
+#include <absl/base/thread_annotations.h>
+
+#include <optional>
 #include <string_view>
-#include <tuple>
+#include <vector>
 
 #include "src/core/fibers.h"
 
 namespace dfly {
 
-typedef uint16_t SlotId;
+using SlotId = uint16_t;
 
 class ClusterConfig {
  public:
-  ClusterConfig();
+  enum class Role {
+    kMaster,
+    kReplica,
+  };
+
+  struct Node {
+    std::string id;
+    std::string ip;
+    uint16_t port = 0;
+    Role role;
+  };
+
+  struct SlotRange {
+    SlotId start = 0;
+    SlotId end = 0;
+  };
+
+  struct ClusterShard {
+    std::vector<SlotRange> slot_ranges;
+    std::vector<Node> nodes;
+  };
+
+  explicit ClusterConfig(std::string_view my_id);
+
   static SlotId KeySlot(std::string_view key);
+
   static bool IsClusterEnabled() {
     return cluster_enabled;
   }
+
   // If the key contains the {...} pattern, return only the part between { and }
   static std::string_view KeyTag(std::string_view key);
 
   // If key is in my slots ownership return true
-  bool IsMySlot(SlotId id);
+  bool IsMySlot(SlotId id) const;
 
- private:
-  void AddSlots();
+  // Returns nodes that own `id`. Empty if owned by me.
+  std::optional<Node> GetNodeForSlot(SlotId id) const;
```

Comment:
You created this function for the MOVE error, right? You will only need tor return the ip and port in this case.
If you save all the config data in SetConfig, as you will need it for later, see my comment on the config member, you can create an array of pair string_view ip and port

Reply:
A `string_view` won't work here I'm afraid, as the config may be changed after we return from this method.
And also - we'll need to return the ID and role for `CLUSTER SHARDS`, so there's no redundant information in returning `Node`

Reply:
The return type of this function can be pair<string,string> while the slots array data can be pair<string_view,sting_view>
I believe you will need to save entire config data not in a vector of slot nodes, but as given in SetConfig command, just vector of shards data. Therefor I suggest having the slots array as pair<string_view,sting_view> which will point to the vector of shards data.

Reply:
I was thinking about keeping only the `config_` array (to be renamed to `slots_`) instead of saving redundant data. Then, upon `CLUSTER SHARDS` or `CLUSTER SLOTS` we could build whatever we need (and it's not like the performance of these command matter). I conceive this to be less error prone (fewer points of failure / mismatches, smaller state to maintain).

Reply:
but you save on the slots_ array only master node

Reply:
Yes, I'm changing that following your feedback above :)